### ISSUE TITLE
Feature request: Configurable word selection

### ISSUE THREAD

**Issue by dolik-rce** [ismail-yilmaz/Bobcat#17]
Currently, double clicking in text selects a word, where word is defined as continuous run of alphanumeric characters or underscore (if I interpret the code in `TerminalCtrl::GetWordSelection` correctly). That is technically correct, but not always much useful.

Sometimes it would be much more pleasant to use, if the selection was smarter and could recognize bigger blocks of text that form single unit of information. Common use cases (for me) would be to use double click to select and then copy:
- file paths (contain also slashes and dots)
- links (can contain a lot of different characters)

I imagine that the exact rules how the double click selection works might be customizable, similar to how Linkifier works. User could configure a regular expresion(s) that would define what is a "word". The default could be the same as current behavior (so something like `\b\B*\b` or `[_[:alnum:]]*`).

Note that this is a low priority feature, I can happily live without it. But I believe it is one of those little things, that could make people happy when using this terminal :-)

**Comment by ismail-yilmaz** [comment]
It would be a nice little feature indeed.

> I imagine that the exact rules how the double click selection works might be customizable, similar to how Linkifier works.

It would be very similar and probably even easier.  It can be easily implemented in directly Bobcat using the existing tool set of TerminalCtrl (without modifying it).

Added to my TODO list.

**Comment by ismail-yilmaz** [comment]
Ok,

I have added a custom word selection filter support in TerminalCtrl: 0dd9b23
I decided to implement it in TerminalCtrl, because it is useful in general, anyway. (Not to mention it was very simple).

Client codes (in this case Bobcat, of course) can now add custom word selection according to their preferences. 
It can be changed on-the-fly, so more than one filter can be easily added. One cool side of the filter support is it is rather a cell block selection function. So, the client code can inspect the properties of the cell and act accordingly. This means that selecting only, say, underlined/bold/italic text, selecting by text color, and other SGR attributes is possible. 

I will add configurable word selection support to Bobcat asap.

**Comment by ismail-yilmaz** [comment]
> Currently, double clicking in text selects a word, where word is defined as continuous run of alphanumeric characters or underscore (if I interpret the code in `TerminalCtrl::GetWordSelection` correctly). That is technically correct, but not always much useful.
> 
> Sometimes it would be much more pleasant to use, if the selection was smarter and could recognize bigger blocks of text that form single unit of information. Common use cases (for me) would be to use double click to select and then copy:
> 
>     * file paths (contain also slashes and dots)
> 
>     * links (can contain a lot of different characters)
> 
> 
> I imagine that the exact rules how the double click selection works might be customizable, similar to how Linkifier works. User could configure a regular expresion(s) that would define what is a "word". The default could be the same as current behavior (so something like `\b\B*\b` or `[_[:alnum:]]*`).
> 
> Note that this is a low priority feature, I can happily live without it. But I believe it is one of those little things, that could make people happy when using this terminal :-)

I have implemented the configurable word selection in Bobcat. See commit 8b3b2a2

It is pretty basic. You can set which extra characters to be treated as a part of a word. E.g. "/.-_".

**Comment by dolik-rce** [comment]
> I have implemented the configurable word selection in Bobcat. See commit 8b3b2a2
> 
> It is pretty basic. You can set which extra characters to be treated as a part of a word. E.g. "/.-_".

Thank you. I have tried and it solves 95% of the problems I've had with the selection before, which is great.

> Using regexp as you suggested is also on my list, but at the moment it is not high on the list.

Maybe I'll get to it sooner than you :-)

**Comment by ismail-yilmaz** [comment]
@dolik-rce 

By the way, while the issue #9 is not resolved yet (changes are still pending), I have added a "selector mode" to TerminalCtrl, which is immediately available in Bobcat (see commit 69efeb1). In selector mode you can navigate & and select (text, word, rectangles) in the terminal's buffer, with keyboard shortcuts. Shortcuts are hard coded at the moment, but they will be configurable soon (I am preparing to make all shortcuts in TerminalCtrl configurable.).

Configurable shortcut:

|Keys | Description |
|---|---|
| Shift+Ctrl+X| Enter selector mode. |

Available hard-coded shortcuts (for the time being):
|Keys| Description|
|---|---|
| Escape | Exit selector mode. |
| Return | Start selection.         |
| Ctrl+T   | Select text. |
| Ctrl+W  | Select word. |
| Ctrl+R | Select rectangle. |
| Up [Arrow key] | Move 1 row up. |
| Down [Arrow key]  | Move 1 row down. |
| Left [Arrow key]  | Move 1 column left. |
| Right [Arrow key] | Move 1 column right. |
| Shift+Left [Arrow key] | Move to the beginning of the row. |
| Shift+Right [Arrow key] | Move to the end of the row. |
| Home | Move to the beginning of the buffer. |
| End | Move to the end of the buffer. |
| Page up | Move one page up. |
| Page down | Move one page down. |

**Comment by ismail-yilmaz** [comment]
> Maybe I'll get to it sooner than you :-)

Good news is that it is now very easy to implement the *smart* selection. I have made the `GetWordSelection` method of TerminalCtrl both protected and virtual. All we need to do is override it and scan the line. (In a similar way I did in Finder or Linkifier.

**Comment by ismail-yilmaz** [comment]
> > I have implemented the configurable word selection in Bobcat. See commit [8b3b2a2](https://github.com/ismail-yilmaz/Bobcat/commit/8b3b2a2b1b74661a53e212197477265fafce7b78)
> > It is pretty basic. You can set which extra characters to be treated as a part of a word. E.g. "/.-_".
> 
> Thank you. I have tried and it solves 95% of the problems I've had with the selection before, which is great.
> 
> > Using regexp as you suggested is also on my list, but at the moment it is not high on the list.
> 
> Maybe I'll get to it sooner than you :-)

"Smart" word selection is implemented. Bobcat will allow two word selection modes: plain & smart. In smart mode, it will first process the regexp pattern, and fallback to plain selection mode on failure. Pattern can be set in emulation config tab.

See: commit 740e9fd

Please check.

**Comment by dolik-rce** [comment]
I tried to test with a very simple configuration for now:
![screen_1709378377](https://github.com/ismail-yilmaz/Bobcat/assets/365613/41a041d6-87e9-4011-856d-61df2130133f)

Interestingly, it sometimes produces inconsistent results. Here is an example:
![screen_1709378527](https://github.com/ismail-yilmaz/Bobcat/assets/365613/2d0ee0b2-7432-4562-bb60-3753c2576e35)

In both terminals, I have clicked on the word "main", but in one case the entire file path is selected and in the other only the word. It happens consistently, even if I just type two random filenames manually, so it is not some artifact of git output.

I did look at the algorithm in `GetWordSelectionByPattern` and I think I understand how it is supposed to works. But I have no idea how it could behave like this :-) Any ideas?

**Comment by dolik-rce** [comment]
> P.s: I could have implemented a list of patterns as we have in linkifier, but IMO that would be an overkill. What do you think?

I agree, it's not really needed. Multiple patterns can be always be joined in two one, if necessary.

**Comment by ismail-yilmaz** [comment]
> I tried to test with a very simple configuration for now: ![screen_1709378377](https://private-user-images.githubusercontent.com/365613/309469420-41a041d6-87e9-4011-856d-61df2130133f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzODIxNDQsIm5iZiI6MTcwOTM4MTg0NCwicGF0aCI6Ii8zNjU2MTMvMzA5NDY5NDIwLTQxYTA0MWQ2LTg3ZTktNDAxMS04NTZkLTYxZGYyMTMwMTMzZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMlQxMjE3MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMGFjNTQyMDFjNmU2YWEwYjlhNjM4YWEyMDM0MTBmYmJjYjdhNzRhZTQ1NjFhNWI5M2Q1MmY0OTc1NjcwZjQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Q37xgyBRWCtph8gi-NIS5tZBeYSMPavxsCTJk-y2Ra0)
> 
> Interestingly, it sometimes produces inconsistent results. Here is an example: ![screen_1709378527](https://private-user-images.githubusercontent.com/365613/309469563-2d0ee0b2-7432-4562-bb60-3753c2576e35.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzODIxNDQsIm5iZiI6MTcwOTM4MTg0NCwicGF0aCI6Ii8zNjU2MTMvMzA5NDY5NTYzLTJkMGVlMGIyLTc0MzItNDU2Mi1iYjYwLTM3NTNjMjU3NmUzNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMlQxMjE3MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMGE5NzE4YzBmOWNlYWNiZTQ4MDIxZDVmYTVlMDlhZmYyY2VlZmI1NDJhOWZjMDViNjc1ZjU2NzM2NzM2NTEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.t3i7R7uQtt7K4JdcPcWOVaX5dn9xMvTPUKk1-MORpD0)
> 
> In both terminals, I have clicked on the word "main", but in one case the entire file path is selected and in the other only the word. It happens consistently, even if I just type two random filenames manually, so it is not some artifact of git output.
> 
> I did look at the algorithm in `GetWordSelectionByPattern` and I think I understand how it is supposed to works. But I have no idea how it could behave like this :-) Any ideas?

Yes, I have an idea.  I'll check it and return asap.

**Comment by ismail-yilmaz** [comment]
Ah! Nevermind, I've -hopefully- spotted the problem. Selection anchor and length should be at least 1 cell wide.
Finder and linkifier doesn't need  that delta. That's what I was overlooking. :)

**Comment by dolik-rce** [comment]
Yes, it seems to work correctly now. I'll give it some more testing over next few days, especially with more complicated patterns and bigger buffers. 

I'm quite curious how will it behave if there is a long buffer with many matches, since the algorithm iterates through all the matches. It is most probably going to be fast enough on reasonable hardware, but I'm not so sure about older machines and things like raspberyPi. It might be fun to optimize it a bit :-)

**Comment by ismail-yilmaz** [comment]
> Yes, it seems to work correctly now. I'll give it some more testing over next few days, especially with more complicated patterns and bigger buffers.
> 
> I'm quite curious how will it behave if there is a long buffer with many matches, since the algorithm iterates through all the matches. It is most probably going to be fast enough on reasonable hardware, but I'm not so sure about older machines and things like raspberyPi. It might be fun to optimize it a bit :-)

I ran a quick test with a file with single line containing 100.000 `Bobcat/Bobcat.h` entries, on a 1024 lines scrollback buffer, it took 2.8 secs. On a 65536 ln buffer, Bobcat just stalled. :)

 First optimization could be limiting the search with the visible buffer.

**Comment by dolik-rce** [comment]
I think it would be reasonable to assume that "word" should not span multiple lines. So searching only from previous \n to next \n should be enough. It might still be a lot of text in some cases (e.g. minified JSON), so taking just some limited neighborhood and expand it iteratively might still be necessary.

**Comment by ismail-yilmaz** [comment]
> I think it would be reasonable to assume that "word" should not span multiple lines. So searching only from previous \n to next \n should be enough. It might still be a lot of text in some cases (e.g. minified JSON), so taking just some limited neighborhood and expand it iteratively might still be necessary.

OK, 

I have tried to improve the situation. First, both linkifier and smart selection are optimized and should be way faster than before. Second, smart selection no longer scans the whole line if the line is longer than approx (>= 2048) characters. It now iterates and expands the line till it hits the upper limit.

(By the way, I have also added minimal shell integration (`OSC 7` protocol) to Bobcat. If your shell is configured correctly via a script, (an example can be found  in wezterm's repo and in other modern terminals), shell will notify Bobcat for directory changes and the new terminal will use that path if it is enabled.)
